File: src/main/java/com/mbayou/kick4k/channels/ChannelsClient.java
package com.mbayou.kick4k.channels;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mbayou.kick4k.KickConfiguration;
import com.mbayou.kick4k.api.ApiClient;
import com.mbayou.kick4k.authorization.AuthorizationClient;

import java.net.http.HttpClient;
import java.util.List;
import java.util.Map;

public class ChannelsClient extends ApiClient {
    public ChannelsClient(HttpClient httpClient, ObjectMapper mapper, KickConfiguration configuration, AuthorizationClient authorization) {
        super(httpClient, mapper, configuration, authorization);
    }

    public Channel getCurrentChannel() {
        List<Channel> channels = this.get(this.configuration.getChannels())
                .send(new TypeReference<>() {});
        return channels.getFirst();
    }

    public List<Channel> getChannels(int... broadcasterUserIds) {
        return this.get(this.configuration.getChannels())
                .queryParams(Map.of("broadcaster_user_id", broadcasterUserIds))
                .send(new TypeReference<>() {});
    }

    public List<Channel> getChannelsByUserIds(List<Integer> broadcasterUserIds) {
        return this.get(this.configuration.getChannels())
                .queryParams(Map.of("broadcaster_user_id", broadcasterUserIds))
                .send(new TypeReference<>() {});
    }

    public Channel getChannel(int broadcasterUserId) {
        return this.getChannels(broadcasterUserId).getFirst();
    }

    public List<Channel> getChannels(String... slugs) {
        return this.get(this.configuration.getChannels())
                .queryParams(Map.of("slug", slugs))
                .send(new TypeReference<>() {});
    }


    public List<Channel> getChannelsBySlugs(List<String> slugs) {
        return this.get(this.configuration.getChannels())
                .queryParams(Map.of("slug", slugs))
                .send(new TypeReference<>() {});
    }

    public Channel getChannel(String slug) {
        return this.getChannels(slug).getFirst();
    }

    public void updateChannel(UpdateChannelRequest request) {
        this.patch(this.configuration.getChannels())
                .body(request)
                .send(new TypeReference<>() {});
    }
}

File: src/main/kotlin/com/mbayou/kick4k/channels/ChannelsClient.kt
package com.mbayou.kick4k.channels

import com.fasterxml.jackson.databind.ObjectMapper
import com.mbayou.kick4k.KickConfiguration
import com.mbayou.kick4k.api.ApiClient
import java.net.http.HttpClient

class ChannelsClient(
    httpClient: HttpClient,
    mapper: ObjectMapper,
    configuration: KickConfiguration,
): ApiClient(httpClient, mapper, configuration) {

    fun getCurrentChannel(accessToken: String): Channel = getChannels(accessToken).first()

    fun getChannels(accessToken: String): List<Channel> {
        return get(configuration.channels)
            .withAccessToken(accessToken)
            .send(responseType<List<Channel>>()) ?: emptyList()
    }

    fun getChannels(accessToken: String, vararg broadcasterUserIds: Int): List<Channel> {
        return get(configuration.channels)
            .queryParams(mapOf("broadcaster_user_id" to broadcasterUserIds))
            .withAccessToken(accessToken)
            .send(responseType<List<Channel>>()) ?: emptyList()
    }

    fun getChannelsByUserIds(accessToken: String, broadcasterUserIds: List<Int>): List<Channel> {
        return get(configuration.channels)
            .queryParams(mapOf("broadcaster_user_id" to broadcasterUserIds))
            .withAccessToken(accessToken)
            .send(responseType<List<Channel>>()) ?: emptyList()
    }

    fun getChannel(accessToken: String, broadcasterUserId: Int): Channel =
        getChannels(accessToken, broadcasterUserId).first()

    fun getChannelsBySlugs(accessToken: String, slugs: List<String>): List<Channel> {
        return get(configuration.channels)
            .queryParams(mapOf("slug" to slugs))
            .withAccessToken(accessToken)
            .send(responseType<List<Channel>>()) ?: emptyList()
    }

    fun getChannels(accessToken: String, vararg slugs: String): List<Channel> {
        return get(configuration.channels)
            .queryParams(mapOf("slug" to slugs))
            .withAccessToken(accessToken)
            .send(responseType<List<Channel>>()) ?: emptyList()
    }

    fun getChannel(accessToken: String, slug: String): Channel = getChannels(accessToken, slug).first()

    fun updateChannel(accessToken: String, request: UpdateChannelRequest) {
        patch(configuration.channels)
            .body(request)
            .withAccessToken(accessToken)
            .send(responseType<Unit>())
    }
}
